README.md
# zxcvbn for Elixir (Rust NIF) [![hex.pm](https://img.shields.io/hexpm/v/zxcvbn_nif.svg?style=flat-square)](https://hex.pm/packages/zxcvbn_nif) [![hexdocs.pm](https://img.shields.io/badge/docs-latest-green.svg?style=flat-square)](https://hexdocs.pm/zxcvbn_nif)

This library is based on the [zxcvbn-rs](https://github.com/shssoichiro/zxcvbn-rs) package written in Rust. It is ported to Elixir as a NIF and offers significantly improved performance over other Elixir libraries.

For more general information, see the original [zxcvbn](https://github.com/dropbox/zxcvbn) package by Dropbox.

## Installation

The package is [available in Hex](https://hex.pm/packages/zxcvbn_nif) and can be installed
by adding `zxcvbn_nif` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:zxcvbn_nif, "~> 1.0"}
  ]
end
```

## Usage

Full documentation is published here: [https://hexdocs.pm/zxcvbn_nif](https://hexdocs.pm/zxcvbn_nif).

Just call `Zxcvbn.run/1` with a password to analyse:

```elixir
iex> Zxcvbn.run("password")
{:ok,
 %{
   calc_time: 1.0761590003967285,
   crack_times_display: %{
     offline_fast_hashing_1e10_per_second: "less than a second",
     offline_slow_hashing_1e4_per_second: "less than a second",
     online_no_throttling_10_per_second: "less than a second",
     online_throttling_100_per_hour: "1 minute"
   },
   feedback: %{
     suggestions: ["Add another word or two. Uncommon words are better.",
      "Reversed words aren't much harder to guess."],
     warning: "This is similar to a commonly used password."
   },
   guesses: 3,
   guesses_log10: 0.47712125471966244,
   score: 0
 }}
```

You can also supply a second optional parameter, to use your own word dictionary:

```elixir
iex> Zxcvbn.run("konnibanwa", ["konni", "banwa"])
{:ok,
 %{
   calc_time: 0.6378800272941589,
   crack_times_display: %{
     offline_fast_hashing_1e10_per_second: "less than a second",
     offline_slow_hashing_1e4_per_second: "1 second",
     online_no_throttling_10_per_second: "25 minutes",
     online_throttling_100_per_hour: "6 days"
   },
   feedback: %{
     suggestions: ["Add another word or two. Uncommon words are better."],
     warning: nil
   },
   guesses: 15000,
   guesses_log10: 4.176091259055681,
   score: 1
 }}
```

To compare, here is the result without using a custom dictionary:

```elixir
iex> Zxcvbn.run("konnibanwa")
{:ok,
 %{
   calc_time: 0.49091002345085144,
   crack_times_display: %{
     offline_fast_hashing_1e10_per_second: "less than a second",
     offline_slow_hashing_1e4_per_second: "10 days",
     online_no_throttling_10_per_second: "29 years",
     online_throttling_100_per_hour: "centuries"
   },
   feedback: %{suggestions: nil, warning: nil},
   guesses: 9460000000,
   guesses_log10: 9.975891136401792,
   score: 3
 }}
```

## Benchmarks
Compared to [zxcvbn](https://hex.pm/packages/zxcvbn) (pure Elixir) package:

```
Operating System: macOS
CPU Information: Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
Number of Available Cores: 4
Available memory: 16 GB
Elixir 1.9.2
Erlang 21.0.3

##### With input Password: elixir nifs are great #####
Name                    ips        average  deviation         median         99th %
zxcvbn-nif           785.01        1.27 ms    ±26.05%        1.21 ms        2.42 ms
zxcvbn-elixir         37.51       26.66 ms    ±15.42%       25.99 ms       50.74 ms

Comparison:
zxcvbn-nif           785.01
zxcvbn-elixir         37.51 - 20.93x slower +25.38 ms

##### With input Password: password #####
Name                    ips        average  deviation         median         99th %
zxcvbn-nif           3.51 K        0.29 ms    ±68.93%        0.27 ms        0.43 ms
zxcvbn-elixir      0.0397 K       25.22 ms    ±12.50%       24.72 ms       43.75 ms

Comparison:
zxcvbn-nif           3.51 K
zxcvbn-elixir      0.0397 K - 88.44x slower +24.93 ms
```

## License

See LICENSE.md file.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.4] - 2021-07-17

- Upgrade dependencies for being able to compile for OTP24
- rustler is bumped to 0.22.0
- zxcvbn is bumped to 2.1.1

## [1.0.2] - 2020-02-25

Fixed compilation error related to otp_app/package name.

## [1.0.1] - 2020-02-25

Fixed error related to NIF folder missing from package.

## [1.0.0] - 2020-02-25

Initial release.


mix.exs
defmodule Zxcvbn.MixProject do
  use Mix.Project

  @version "1.0.4"

  def project do
    [
      app: :zxcvbn_nif,
      version: @version,
      elixir: "~> 1.9",
      name: "Zxcvbn",
      description: "zxcvbn for Elixir using Rust NIF.",
      deps: [
        {:rustler, "~> 0.22.0"},
        {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
        {:benchee, "~> 1.0", only: :dev}
        # {:zxcvbn, "~> 0.1.3", only: :dev}
      ],
      package: [
        name: "zxcvbn_nif",
        maintainers: ["Svilen Gospodinov <webmaster@s2g.io>"],
        licenses: ["MIT"],
        links: %{Github: "https://github.com/svileng/zxcvbn"},
        files: ["lib", "native", "mix.exs", "README.md", "LICENSE.md"]
      ],
      docs: [
        main: "zxcvbn",
        canonical: "http://hexdocs.pm/zxcvbn",
        source_url: "https://github.com/svileng/zxcvbn",
        source_ref: @version
      ]
    ]
  end
end


native/zxcvbn_wrapper/src/lib.rs
#[macro_use]
extern crate rustler;

use rustler::{Encoder, Env, ListIterator, Term};
use zxcvbn::{Entropy, ZxcvbnError};

mod atoms {
    atoms! {
        ok,
        error,
        nil,
        score,
        guesses,
        guesses_log10,
        calc_time,
        crack_times_display,
        online_throttling_100_per_hour,
        online_no_throttling_10_per_second,
        offline_slow_hashing_1e4_per_second,
        offline_fast_hashing_1e10_per_second,
        feedback,
        warning,
        suggestions,
    }
}

rustler::init!("Elixir.Zxcvbn", [run]);

struct EntropyWrap {
    entropy: Entropy,
}

impl Encoder for EntropyWrap {
    fn encode<'a>(&self, env: Env<'a>) -> Term<'a> {
        let entropy = &self.entropy;

        let crack_times = ::rustler::types::map::map_new(env)
            .map_put(
                atoms::offline_fast_hashing_1e10_per_second().encode(env),
                entropy
                    .crack_times()
                    .offline_fast_hashing_1e10_per_second()
                    .to_string()
                    .encode(env),
            )
            .ok()
            .unwrap()
            .map_put(
                atoms::offline_slow_hashing_1e4_per_second().encode(env),
                entropy
                    .crack_times()
                    .offline_slow_hashing_1e4_per_second()
                    .to_string()
                    .encode(env),
            )
            .ok()
            .unwrap()
            .map_put(
                atoms::online_throttling_100_per_hour().encode(env),
                entropy
                    .crack_times()
                    .online_throttling_100_per_hour()
                    .to_string()
                    .encode(env),
            )
            .ok()
            .unwrap()
            .map_put(
                atoms::online_no_throttling_10_per_second().encode(env),
                entropy
                    .crack_times()
                    .online_no_throttling_10_per_second()
                    .to_string()
                    .encode(env),
            )
            .ok()
            .unwrap();

        let mut feedback = ::rustler::types::map::map_new(env);
        feedback = match entropy.feedback().clone() {
            Some(feedback_value) => feedback
                .map_put(
                    atoms::warning().encode(env),
                    match feedback_value.warning() {
                        Some(warning_value) => warning_value.to_string().encode(env),
                        None => atoms::nil().encode(env),
                    },
                )
                .ok()
                .unwrap()
                .map_put(
                    atoms::suggestions().encode(env),
                    match feedback_value.suggestions() {
                        [] => atoms::nil().encode(env),
                        vector => vector
                            .iter()
                            .map(|s| s.to_string())
                            .collect::<Vec<_>>()
                            .encode(env),
                    },
                )
                .ok()
                .unwrap(),
            None => feedback
                .map_put(atoms::warning().encode(env), atoms::nil().encode(env))
                .ok()
                .unwrap()
                .map_put(atoms::suggestions().encode(env), atoms::nil().encode(env))
                .ok()
                .unwrap(),
        };

        let result = ::rustler::types::map::map_new(env)
            .map_put(atoms::score().encode(env), entropy.score().encode(env))
            .ok()
            .unwrap()
            .map_put(atoms::guesses().encode(env), entropy.guesses().encode(env))
            .ok()
            .unwrap()
            .map_put(
                atoms::guesses_log10().encode(env),
                entropy.guesses_log10().encode(env),
            )
            .ok()
            .unwrap()
            .map_put(atoms::crack_times_display().encode(env), crack_times)
            .ok()
            .unwrap()
            .map_put(
                atoms::calc_time().encode(env),
                // `calculation_time()` is a `Duration`
                // in seconds. We're expecting ms so
                // we multiply that by 1000.
                entropy
                    .calculation_time()
                    .mul_f32(1000.0)
                    .as_secs_f32()
                    .encode(env),
            )
            .ok()
            .unwrap()
            .map_put(atoms::feedback().encode(env), feedback)
            .ok()
            .unwrap();

        result
    }
}

enum RunReturn {
    Ok(Entropy),
    Error(&'static str),
}

impl Encoder for RunReturn {
    fn encode<'a>(&self, env: Env<'a>) -> Term<'a> {
        match self {
            RunReturn::Ok(entropy) => (
                atoms::ok(),
                EntropyWrap {
                    entropy: entropy.to_owned(),
                },
            )
                .encode(env),
            RunReturn::Error(reason) => (atoms::error(), reason).encode(env),
        }
    }
}

#[rustler::nif(schedule = "DirtyCpu", name = "run_nif")]
fn run(password: &str, inputs: ListIterator) -> RunReturn {
    let user_inputs = inputs.map(|i| i.decode().unwrap_or("")).collect::<Vec<_>>();

    match zxcvbn::zxcvbn(password, &user_inputs) {
        Ok(entropy) => RunReturn::Ok(entropy),
        Err(ZxcvbnError::BlankPassword) => RunReturn::Error("blank_password"),
        Err(ZxcvbnError::DurationOutOfRange) => RunReturn::Error("duration_out_of_range"),
    }
}


lib/zxcvbn.ex
defmodule Zxcvbn do
  @moduledoc """
  To use `Zxcvbn` simply call `run/1` with a password string to analyse:

      iex> Zxcvbn.run("password")
      {:ok,
      %{
        calc_time: 1.0761590003967285,
        crack_times_display: %{
          offline_fast_hashing_1e10_per_second: "less than a second",
          offline_slow_hashing_1e4_per_second: "less than a second",
          online_no_throttling_10_per_second: "less than a second",
          online_throttling_100_per_hour: "1 minute"
        },
        feedback: %{
          suggestions: ["Add another word or two. Uncommon words are better.",
            "Reversed words aren't much harder to guess."],
          warning: "This is similar to a commonly used password."
        },
        guesses: 3,
        guesses_log10: 0.47712125471966244,
        score: 0
      }}

  The result will be either `{:ok, result}`, `{:error, "blank_password"}` or `{:error, "duration_out_of_range"}`.

  The keys found in a successful response are documented in the [original `zxcvbn` repository](https://github.com/dropbox/zxcvbn) by Dropbox. Note that not all keys from the original library are implemented in `zxcvbn_nif`.

  A second optional parameter `user_inputs` is supported for custom word dictionaries, for example:

      iex> Zxcvbn.run("konnibanwa", ["konni", "banwa"])

  As described in the original `zxcvbn` documentation, the custom dictionaries should be a list of strings.
  """
  use Rustler, otp_app: :zxcvbn_nif, crate: "zxcvbn_wrapper"

  defp run_nif(_, _), do: :erlang.nif_error(:nif_not_loaded)

  @spec run(String.t()) :: {:ok, map} | {:error, String.t()}
  def run(""), do: {:error, "blank_password"}
  def run(password), do: run_nif(password, [])

  @spec run(String.t(), list(String.t())) :: {:ok, map} | {:error, String.t()}
  def run("", _), do: {:error, "blank_password"}

  def run(password, user_input) when is_list(user_input),
    do: run_nif(password, user_input)
end
